#' Get Gene Nearby Range
#'
#' Given a gene name, return its nearby region GRanges.
#'
#' @param gtf a GRanges imported by rtracklayer from GTF file.
#' @param gene name of a gene.
#' @param upstream upstream bp.
#' @param downstream downstream bp.
#'
#' @export
get_gene_range <- function(gtf, gene, upstream = 2000, downstream = 2000) {
  x <- gtf[gtf$gene_id == gene, ] |> range()
  GenomicRanges::start(x) <- GenomicRanges::start(x) - upstream
  GenomicRanges::end(x) <- GenomicRanges::end(x) + downstream
  GenomicRanges::strand(x) <- "*"
  return(x)
}


#' Plot Gene Track with Primary Transcripts
#'
#' @param gtf A GRanges, generated by rtracklayer::import from a GTF file in ensembl format.
#' @param gene_range A GRanges, specify a region of gtf to plot gene track.
#' @param padding A integer, it is the padding of genes' upstream and downstream so that its name not overlapping.
#' @param show_gene_id Boolean, Whether to show gene_id of transcripts.
#' @param show_arrow Boolean, Whether to show arrow indicating orientation of transcripts.
#' @param remove A vector of gene_id, remove transcripts indicated.
#' @param keep_only A vector of gene_id, keep transcripts indicated.
#'
#' @export
#'
plot_gene_track <- function(gtf, gene_range, padding = 0, show_gene_id = TRUE,
                            show_arrow = TRUE, remove = FALSE, keep_only = FALSE) {
  # Given a GRanges of one gene, find its primary transcript, and return its GRanges.
  get_primary_tx <- function(gene_gtf, by = "cds") {
    tx_list <- split(gene_gtf, gene_gtf$transcript_id)

    if (by == "cds") {
      x <- sapply(tx_list, function(tx) {
        subset(tx, tx$type == "CDS") |>
          IRanges::width() |>
          sum()
      }) |>
        sort(decreasing = TRUE)
      primary_tx_id <- names(x)[[1]]
    }

    primary_tx <- tx_list[[primary_tx_id]]
    return(primary_tx)
  }

  # Given a list of GRanges, distribute their ranges into different tracks
  get_track_list <- function(grl) {
    range_list <- lapply(grl, range) |> lapply(BiocGenerics::unstrand)
    track_list <- list(setNames(list(range_list[[1]]), names(range_list)[[1]]))

    if (length(range_list) == 1) {
      return(track_list)
    }

    for (i in 2:length(range_list)) {
      this_range <- range_list[[i]]
      start(this_range) <- start(this_range) - padding
      end(this_range) <- end(this_range) + padding
      this_range_name <- names(range_list)[[i]]

      j <- 0
      overlap <- TRUE
      while (overlap == TRUE) {
        j <- j + 1
        if (j > length(track_list)) {
          track_list[[j]] <- list()
          overlap <- FALSE
        } else {
          overlap <- any(sapply(track_list[[j]], function(x) x %over% this_range))
        }
      }
      track_list[[j]][[this_range_name]] <- this_range
    }
    return(track_list)
  }


  add_track_num <- function(grl) {
    track_list <- get_track_list(grl)

    track_num_list <- rev(seq(2, 2 * length(track_list), length.out = length(track_list)))

    for (i in seq_len(length(track_list))) {
      for (name in names(track_list[[i]])) {
        grl[[name]]$track_num <- track_num_list[[i]]
      }
    }
    return(grl)
  }

  get_arrow_pos <- function(plot_data, width = 500) {
    cds <- subset(plot_data, type == "CDS")
    if (all(cds$strand == "+")) {
      start <- min(cds$start)
      end <- start + width
    } else if (all(cds$strand == "-")) {
      start <- max(cds$end)
      end <- start - width
    }
    return(c(start = start, end = end, track_num = cds$track_num[[1]]))
  }



  gene_in_range <- unique(subsetByOverlaps(gtf, gene_range)$gene_id)

  if (remove != FALSE) {
    gene_in_range <- gene_in_range[!gene_in_range %in% remove]
  }

  if (keep_only != FALSE) {
    gene_in_range <- keep_only
  }

  range_gr <- subset(gtf, gene_id %in% gene_in_range)

  plot_data <- S4Vectors::split(range_gr, ~gene_id) |>
    lapply(get_primary_tx) |>
    add_track_num() |>
    lapply(as.data.frame) |>
    do.call(what = rbind, args = _)

  track_plot <- ggplot() +
    geom_rect(data = subset(plot_data, type == "transcript"), mapping = aes(xmin = start, xmax = end, ymin = track_num - 0.05, ymax = track_num + 0.05), fill = "black", color = "black") +
    geom_rect(data = subset(plot_data, type == "exon"), mapping = aes(xmin = start, xmax = end, ymin = track_num - 0.5, ymax = track_num + 0.5), fill = "white", color = "black") +
    geom_rect(data = subset(plot_data, type == "CDS"), mapping = aes(xmin = start, xmax = end, ymin = track_num - 0.5, ymax = track_num + 0.5), fill = "black", color = "black") +
    coord_cartesian(xlim = c(start(gene_range), end(gene_range)), ylim = c(min(plot_data$track_num) - 2, max(plot_data$track_num) + 2)) +
    scale_x_continuous(expand = c(0, 1)) +
    scale_y_continuous(breaks = NULL, labels = NULL) +
    labs(x = NULL, y = NULL) +
    theme_classic()

  if (show_gene_id) {
    track_plot <- track_plot +
      geom_text(data = subset(plot_data, type == "transcript"), mapping = aes(x = (start + end) / 2, y = track_num - 1, label = gene_id))
  }

  if (show_arrow) {
    arrow_data <- subset(plot_data, type == "CDS") |>
      S4Vectors::split(~transcript_id) |>
      lapply(get_arrow_pos) |>
      do.call(what = rbind, args = _) |>
      as.data.frame()

    track_plot <- track_plot +
      geom_segment(data = arrow_data, mapping = aes(x = start, y = track_num + 1, xend = end, yend = track_num + 1), arrow = grid::arrow(angle = 45, length = grid::unit(0.1, "inches"))) +
      geom_segment(data = arrow_data, mapping = aes(x = start, y = track_num, xend = start, yend = track_num + 1))
  }

  return(track_plot)
}
